lib/src/model.dart

```diff
@@ -2243,32 +2476,138 @@ class Operator extends Method {
   String get typeName => 'operator';
 }
 
+// The kinds of warnings that can be displayed when documenting a package.
+enum PackageWarning {
+  ambiguousReexport,
+  noCanonicalFound,
+  noLibraryLevelDocs,
+}
+
 class Package implements Nameable, Documentable {
+  // Library objects serving as entry points for documentation.
   final List<Library> _libraries = [];
+  // All library objects related to this package; a superset of _libraries.
+  final Map<LibraryElement, Library> _all_libraries = new Map();
+  // All ModelElements constructed for this package; a superset  of allModelElements.
+  final Map<Tuple3<Element, Library, Class>, ModelElement> _all_model_elements =
+      new Map();
+
+  /// Map of Class.href to a list of classes implementing that class
+  final Map<String, List<Class>> _implementors = new Map();
+
+  final Map<Element, Set<PackageWarning>> _displayedWarnings = new Map();
+
   final PackageMeta packageMeta;
-  final Map<String, Library> elementLibaryMap = {};
+
+  final Map<Element, Library> _elementToLibrary = {};
   String _docsAsHtml;
   final Map<String, String> _macros = {};
+  bool allLibrariesAdded = false;
 
   Package(Iterable<LibraryElement> libraryElements, this.packageMeta) {
+    assert(_all_model_elements.isEmpty);
+    assert(_all_libraries.isEmpty);
     libraryElements.forEach((element) {
       // add only if the element should be included in the public api
       if (isPublic(element)) {
-        var lib = new Library(element, this);
-        Library._libraryMap.putIfAbsent(lib.name, () => lib);
-        elementLibaryMap.putIfAbsent('${lib.kind}.${lib.name}', () => lib);
+        var lib = new Library._(element, this);
         _libraries.add(lib);
+        _all_libraries[element] = lib;
+        assert(!_elementToLibrary.containsKey(lib.element));
+        _elementToLibrary[element] = lib;
       }
     });
 
+    _libraries.sort((a, b) => compareNatural(a.name, b.name));
+    allLibrariesAdded = true;
     _libraries.forEach((library) {
       library._allClasses.forEach(_addToImplementors);
     });
 
-    _libraries.sort();
     _implementors.values.forEach((l) => l.sort());
   }
 
+  void warn(ModelElement modelElement, PackageWarning kind, [String message]) {
```

Comment:
Not for now, but when doing future work in there area - we'll likely want to be able to print out source locations for where these warnings are occurring.